src/ui/VirtualFileRenderer/VirtualFileRenderer.tsx

```diff
@@ -0,0 +1,323 @@
+  }
+
+  // TODO - finalize this logic
+  // scroll to line on the initial render
+  useEffect(() => {
+    if (initialRender.current) {
+      initialRender.current = false
+      const index = parseInt(location.hash.slice(2), 10)
+      virtualizer.scrollToIndex(index)
+    }
+  }, [location.hash, virtualizer])
+
+  return (
+    <div className="flex flex-1" ref={wrapperRef}>
+      {/* this div contains the line numbers */}
+      <div className="relative z-[2] h-full w-[82px] min-w-[82px] pr-[10px]">
+        {virtualizer.getVirtualItems().map((item) => (
+          <div
+            ref={virtualizer.measureElement}
+            key={item.index}
+            data-index={item.index}
+            style={{
+              height: `${item.size}px`,
+              transform: `translateY(${
+                item.start - virtualizer.options.scrollMargin
+              }px)`,
+            }}
+            className={cn(
+              `absolute left-0 top-0 w-full border-r border-ds-gray-tertiary bg-white px-4 text-right text-ds-gray-quaternary hover:cursor-pointer hover:text-black`,
+              location.hash === `#L${item.index}` &&
+                'font-semibold text-ds-gray-quinary',
+              coverage[item.index] === 'H' && 'bg-ds-coverage-covered',
+              coverage[item.index] === 'M' &&
+                'bg-ds-coverage-uncovered after:absolute after:inset-y-0 after:right-0 after:border-r-2 after:border-ds-primary-red',
+              coverage[item.index] === 'P' &&
+                'bg-ds-coverage-partial after:absolute after:inset-y-0 after:right-0 after:border-r-2 after:border-dotted after:border-ds-primary-yellow'
+            )}
+            onClick={() => {
+              if (location.hash === `#L${item.index}`) {
+                location.hash = ''
+              } else {
+                location.hash = `#L${item.index}`
+              }
+              history.push(location)
+            }}
+          >
+            {location.hash === `#L${item.index}` ? '#' : null}
+            {item.index}
+          </div>
+        ))}
+      </div>
+      {/* this div contains the actual code lines */}
+      <div
+        // @ts-expect-error - TODO - Update inert when inert is available in React 19
+        inert=""
+        className="pointer-events-none size-full"
+      >
+        {virtualizer.getVirtualItems().map((item) => {
+          // get any specific things from code highlighting library for this given line
+          const { style: lineStyle } = getLineProps({
+            // casting this cause it is guaranteed to be present in the array
+            line: tokens[item.index] as Token[],
+            key: item.index,
+          })
+
+          return (
+            <div
+              ref={virtualizer.measureElement}
+              key={item.index}
+              data-index={item.index}
+              style={{
+                width: wrapperRef?.current
+                  ? wrapperRef.current.clientWidth
```

Comment:
This is going to cause layout thrashing and potential perf issues as it will read the clientWidth of the wrapper for each rendered row.

Reply:
Would setting this to a const under the ref definition work? or would i need to set it in something like a useEffect and set a piece of state?

Reply:
This seems to have worked fine, so now it's only getting rendered in the parent component, should stop any issues.

Reply:
The way I would usually handle this is by attaching a resize observer and storing the value. That way you get the value once when it renders and store it in state (or assign it directly via css var), then keep it in sync as it resizes

Reply:
This size is static, its the max width of the textarea, so it won't change even on resize.

Reply:
So this value is actually the width of the text area so it's fairly static even with resizing the window, so I think we can probably skip the resize observer in this instance and just leave it be, lmk what you think.

Reply:
I would add the resize observer to be safe. It is safer, more correct and might save you some time from finding misalignment issues in the future.

Reply:
alrighty done